android/engine/src/main/java/org/smartregister/fhircore/engine/rulesengine/RulesFactory.kt

```diff
@@ -774,6 +778,39 @@ constructor(
         }
       }
     }
+
+    fun getResourceByReference(
+      resourceReference: String?,
+    ): Resource? {
+      if (resourceReference.isNullOrEmpty()) {
+        return null
+      }
+
+      return runBlocking(dispatcherProvider.io()) {
```

Comment:
If you call runBlocking(Dispatchers.IO) from the main thread, the main thread will be blocked, but the coroutine itself will run on a thread from the Dispatchers.IO pool. Use `CoroutineScope` instead.

Reply:
Isn't `CoroutineScope` ideal when we don't need to return a value? CoroutineScope may return earlier before we get return value.

Reply:
Not really, CoroutineScope. You can return using `async` function of the CoroutineScope. 

i.e.
```kotlin 
val coroutineScope = CoroutineScope(dispatcherProvider.io())

val result = coroutineScope.async {
// Your logic that returns a value
}.await()

```
`async` will defer the returned value when the `await` function is called.